### PR TITLE
give the QueryBuilder to callback function in callIfNecessary

### PR DESCRIPTION
**Pull request by mathroc**
see #88 

I have not found a way to get access to the QueryBuilder yet from a orderby callback spec. maybe some sleep will help, meanwhile. I could complete my task with this patch to the QueryBuilder

### CODE REVIEW

File: packages/graphile-build-pg/src/QueryBuilder.js [hunk 1]
```diff
@@ -7,17 +7,23 @@ const isDev = ["test", "development"].indexOf(process.env.NODE_ENV) >= 0;
...
 }
 
-function callIfNecessaryArray<T>(o: Array<Gen<T> | T>): Array<T> {
+function callIfNecessaryArray<T>(
+  o: Array<Gen<T> | T>,
+  context: { queryBuilder: QueryBuilder }
+): Array<T> {
   if (Array.isArray(o)) {
-    return o.map(callIfNecessary);
+    return o.map(callIfNecessary, context);
```
Comment (benjie) [inline packages/graphile-build-pg/src/QueryBuilder.js]: Rather: `return o.map(v => callIfNecessary(v, context));`

File: packages/graphile-build-pg/src/QueryBuilder.js [hunk 2]
```diff
@@ -495,38 +501,51 @@ class QueryBuilder {
...
     } else if (type === "limit") {
-      this.compiledData[type] = callIfNecessary(this.data[type]);
+      this.compiledData[type] = callIfNecessary(this.data[type], {
+        queryBuilder: this,
+      });
     } else if (type === "offset") {
-      this.compiledData[type] = callIfNecessary(this.data[type]);
+      this.compiledData[type] = callIfNecessary(this.data[type], {
+        queryBuilder: this,
+      });
```
Comment (benjie) [inline packages/graphile-build-pg/src/QueryBuilder.js]: Let's DRY this up by putting something like:

```js
const getContext = () => ({
  queryBuilder: this,
})
```
at the top of the `lock` method and then calling it at each place instead:

```js
this.compiledData[type] = callIfNecessary(this.data[type], getContext());
```

File: packages/graphile-build-pg/src/QueryBuilder.js [hunk 3]
```diff
@@ -7,17 +7,23 @@ const isDev = ["test", "development"].indexOf(process.env.NODE_ENV) >= 0;
...
 type Gen<T> = () => T;
 
-function callIfNecessary<T>(o: Gen<T> | T): T {
+function callIfNecessary<T>(
+  o: Gen<T> | T,
+  context: { queryBuilder: QueryBuilder }
+): T {
   if (typeof o === "function") {
-    return o();
+    return o(context);
```
Comment (benjie) [inline packages/graphile-build-pg/src/QueryBuilder.js]: I think we need to modify the type of `Gen<T>` to enable this. Probably just

```js
type GenContext = {
  queryBuilder: QueryBuilder,
};
type Gen<T> = (context: GenContext) => T;

function callIfNecessary<T>(
  o: Gen<T> | T,
  context: GenContext
): T {
...

function callIfNecessaryArray<T>(
  o: Array<Gen<T> | T>,
  context: GenContext
): Array<T> {
```

**Comment by mathroc** [comment]
after some more research I’ve not found another solution for #88. and I don’t think there is one because the query builder has probably not been created before the call to the callback function. 
so I’ve updated the PR

File: packages/graphile-build-pg/src/QueryBuilder.js [hunk 4]
```diff
@@ -5,19 +5,25 @@ import isSafeInteger from "lodash/isSafeInteger";
 
 const isDev = ["test", "development"].indexOf(process.env.NODE_ENV) >= 0;
 
-type Gen<T> = () => T;
+type Gen<T> = (context: GenContext) => T;
```
Comment (benjie) [inline packages/graphile-build-pg/src/QueryBuilder.js]: You omitted the definition of `GenContext`?

**Comment by mathroc** [comment]
tests are green !

File: packages/graphile-build-pg/src/QueryBuilder.js [hunk 5]
```diff
@@ -5,19 +5,28 @@ import isSafeInteger from "lodash/isSafeInteger";
...
-type Gen<T> = () => T;
+type GenContext = {
+  queryBuilder: QueryBuilder,
+};
+type Gen<T> = (context: GenContext) => T;
 
-function callIfNecessary<T>(o: Gen<T> | T): T {
+function callIfNecessary<T>(
+  o: Gen<T> | T,
+  context: { queryBuilder: QueryBuilder }
```
Comment (benjie) [inline packages/graphile-build-pg/src/QueryBuilder.js]: This should be context: GenContext I think?

File: packages/graphile-build-pg/src/QueryBuilder.js [hunk 6]
```diff
@@ -5,19 +5,28 @@ import isSafeInteger from "lodash/isSafeInteger";
...
+    return o(context);
   } else {
     return o;
   }
 }
 
-function callIfNecessaryArray<T>(o: Array<Gen<T> | T>): Array<T> {
+function callIfNecessaryArray<T>(
+  o: Array<Gen<T> | T>,
+  context: { queryBuilder: QueryBuilder }
```
Comment (benjie) [inline packages/graphile-build-pg/src/QueryBuilder.js]: Ditto

**Comment by benjie** [comment]
Hah; I wish prettier would run when editing via GitHub!

/me checks out the branch

**Comment by benjie** [comment]
Merged; thanks! 🙏 

I've got a few other bugs to fix and I should be rolling it into a release late next week most likely.

**Pull request merged**